pkg/operator/ceph/cluster/watcher.go

```diff
@@ -130,6 +145,333 @@ func (c *clientCluster) onK8sNode(ctx context.Context, object runtime.Object) bo
+// 		Entity         entity         `json:"entity"`
+// 		ClientMetadata clientMetadata `json:"client_metadata"`
+// 	}
+
+// 	var data []cephFSData
+// 	err := json.Unmarshal(output, &data)
+// 	if err != nil {
+// 		return []string{}, err
+// 	}
+
+// 	for _, d := range data {
+// 		if cephFSPV.Spec.CSI.VolumeAttributes["subvolumePath"] == d.ClientMetadata.Root {
+// 			logger.Infof("cephfs mds client ips to fence %v", d.Entity.Addr)
+// 			return []string{d.Entity.Addr.Addr + "/32"}, nil
+// 		}
+// 	}
+
+// 	return []string{}, nil
+// }
+
+func (c *clientCluster) createNetworkFence(ctx context.Context, pv corev1.PersistentVolume, node *corev1.Node, cluster *cephv1.CephCluster, cidr []string) error {
+	logger.Warningf("Blocking node IP %s", cidr)
+
+	secretName := pv.Annotations["volume.kubernetes.io/provisioner-deletion-secret-name"]
+	secretNameSpace := pv.Annotations["volume.kubernetes.io/provisioner-deletion-secret-namespace"]
+	if secretName == "" || secretNameSpace == "" {
+		storageClass, err := c.context.Clientset.StorageV1().StorageClasses().Get(ctx, pv.Spec.StorageClassName, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+		secretName = storageClass.Parameters["csi.storage.k8s.io/provisioner-secret-name"]
+		secretNameSpace = storageClass.Parameters["csi.storage.k8s.io/provisioner-secret-namespace"]
+	}
+
+	networkFence := &addonsv1alpha1.NetworkFence{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      node.Name,
+			Namespace: cluster.Namespace,
+			OwnerReferences: []metav1.OwnerReference{
+				*metav1.NewControllerRef(cluster, cephv1.SchemeGroupVersion.WithKind("CephCluster")),
+			},
+		},
+		Spec: addonsv1alpha1.NetworkFenceSpec{
+			Driver:     pv.Spec.CSI.Driver,
+			FenceState: addonsv1alpha1.Fenced,
+			Secret: addonsv1alpha1.SecretSpec{
+				Name:      secretName,
+				Namespace: secretNameSpace,
+			},
+			Cidrs: cidr,
+			Parameters: map[string]string{
+				"clusterID": pv.Spec.CSI.VolumeAttributes["clusterID"],
+			},
+		},
+	}
+
+	err := c.client.Create(ctx, networkFence)
+	if err != nil && !errors.IsAlreadyExists(err) {
+		return err
+	}
+
+	logger.Infof("successfully created network fence CR %s", node.Name)
+
+	return nil
+}
+
+func (c *clientCluster) unfenceAndDeleteNetworkFence(ctx context.Context, node corev1.Node, cluster *cephv1.CephCluster) error {
+	logger.Infof("unfencing node %q", node.Name)
+	networkFence := &addonsv1alpha1.NetworkFence{}
+	err := c.client.Get(ctx, types.NamespacedName{Name: node.Name, Namespace: cluster.Namespace}, networkFence)
+	if err != nil && !errors.IsNotFound(err) {
+		return err
+	} else if errors.IsNotFound(err) {
+		return nil
+	}
+
+	// Unfencing is required to unblock the node and then delete the network fence CR
+	networkFence.Spec.FenceState = addonsv1alpha1.Unfenced
+	err = c.client.Update(ctx, networkFence)
+	if err != nil {
+		logger.Errorf("failed to unFence network fence CR. %v", err)
+		return err
+	}
+
+	err = c.client.Delete(ctx, networkFence)
```

Comment:
If the csi addons controller is down when the update is made, it won't respond to that event before the CR is just deleted. Seems like we just need to delete the CR after we have confirmed the status on the unfencing has been handled. Or maybe we just leave it in the unfenced state and not delete it. Then if the node needs to be fenced in the future, we would update the CR again. This is a design change that could be a separate PR.

Reply:
I'm not very sure about scenario, let's discuss more

Reply:
Yes i would also suggest ensuring the node is unfenced by checking the CR status before deleting the CR.

Reply:
added the code to check that